synapse/util/caches/lrucache.py

```diff
@@ -219,20 +482,34 @@ def __init__(
         # this is exposed for access from outside this class
         self.metrics = metrics
 
-        list_root = _Node(None, None, None, None)
-        list_root.next_node = list_root
-        list_root.prev_node = list_root
+        # We create a single weakref to self here, so that all `_Node` can share
+        # a single reference (as weakrefs are surprisingly large).
```

Comment:
Ooooh, interesting. I did see memory usage go down when removing repeated weakrefs, but I guess that was actually because we kept adding new finalizers every time we took a new ref (originally I added a new finalizer to call `_Node.remove_from_list` for each node).